### PR TITLE
FileUtils: Add a recursive `chmod_r` and `chown_r`

### PR DESCRIPTION
**Pull request by j8r**
Implements a recursive `chown` and `chmod` that applies to all subfiles and subdirectories at a given path.

### CODE REVIEW

**Comment by straight-shoota** [comment]
I'm not sure if these dedicated methods make sense... it's just quite easy to get this done with the non-recursive variants.
With the new glob implementation from #5179 the overhead is really negligible.
```crystal
Dir["path/**"] { |path| File.chmod(path, mode) }
```

File: spec/std/file_utils_spec.cr [hunk 1]
```diff
@@ -183,6 +183,44 @@ describe "FileUtils" do
...
+        File.stat(File.join(path, "a/b")).perm.should eq(0o775)
+      ensure
+        FileUtils.rm_r(path)
+      end
+    end
+  end
+
+  describe "chown_r" do
+    it "change the owner of the directory recursively" do
+      data_path = File.join(__DIR__, "data")
```
Comment (straight-shoota) [inline spec/std/file_utils_spec.cr]: If you don't plan on actually running the method, there is no point in setting up the directories.

**Comment by j8r** [comment]
@straight-shoota I'm thinking now if it stiil makes sense to have `FileUtils` in the stdlib - there are some non essentials methods like aliases to `Dir` and `File`, and others that can be done via the new glob implementation. But if I undestand correctly the implementation, we can't modify the root directory itself this way but the sub directories/files?

**Comment by lachlan** [comment]
I realise there is already a precedence for using different method names for different modes in `FileUtils`, such as `rm_r`, but could we consider using either method overloads or default values which I feel would be a cleaner API.

So instead of:

```crystal
def chmod_r(path, mode : Int)
  ...
end
```

We'd have:

```crystal
def chmod(path, mode : Int, recurse : Bool = false)
  ...
end
```

I'd also prefer methods to be named for the conceptual action they take, rather than named after the historic way to take this action in an operating system's shell, and to use words rather than abbreviations or acronyms. So instead of `rm`, why not `remove` or `delete`? Instead of `chmod`, why not `grant` or something?

More generally, the use of `_suffix` method naming in Crystal to provide different ways of taking the same action does not seem idiomatic. Ruby may do this due to the lack of method overloading, however Crystal does not have this problem. 

I feel the method name should equal the action taken, it should not equal the action taken + options applied to the action.

**Comment by straight-shoota** [comment]
@j8r Yes, the root folder must be handled separtely, but I'd still prefer that over an additional method.

Maybe it would make sense to have something like `Dir.walk_tree(path)` which yields each item in that path to avoid globbing.

**Comment by j8r** [comment]
I agree with you @lachlan for the `recurse` Boolean, underscores are not elegant here. We can merge most of the methods of the `FileUtils` Object to the `File` one. `mkdir_p`, `rm_r` and `cp_r` can be also be changed this way.

In an other hand I disagree to replace all LibC references by new names, specially `chmod` which use an octal number like `0o750`, or `chown` that use uids and gids. This are very specific arguments.

For `mkdir(_p)`, which take only one argument, and optionally an other for the permission mode, why not:
``` crystal
Dir.create("dir/", recurse: true)
Dir.delete("dir/")
```

`rm` doesn't exist as is in the LibC, so replacing it by `delete` is good!

I very like the `recurse` - this translates an Unix argument to a Crystal one.

**Comment by straight-shoota** [comment]
Please let's not discuss these advanced ideas in this PR, open a new issue for that.

**Comment by j8r** [comment]
Sorry, I open an issue

**Comment by j8r** [comment]
Issue opened at #5231

**Comment by j8r** [comment]
I close this PR. `FileUtils`, `File` and `Dir` will certainly be cleaned in the future and then a recursive approach will be implemented for most commands in the future.

**Pull request closed without merging**